test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+    it('sets the auth token if passed', () => {
+      const authToken = 'asdf';
+      const clientId = '1234';
+      const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
+
+      new Auth({
+        clientId,
+        host: 'http://localhost:1738',
+        authToken,
+        messenger: new Messenger({}),
+      });
+      expect(storageSpy).toHaveBeenCalledWith('wally:1234:token', authToken);
+
+      storageSpy.mockRestore();
+    });
+  });
+
+  describe('login', () => {
+    const clientId = '1234';
+    const host = 'http://localhost:1738';
+    let auth: Auth;
+
+    beforeEach(() => {
+      auth = new Auth({
+        clientId,
+        host,
+        authToken: 'asdf',
+        messenger: new Messenger({
+          sharedWorkerUrl: '/worker.js',
+        }),
+      });
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+    });
+
+    it('rejects if already logging in', () => {
+      auth.isLoggingIn = true;
+      return expect(auth.login()).rejects.toBeDefined();
+    });
+
+    it('shows console error if no client id set', () => {
+      const consoleSpy = jest.spyOn(console, 'error');
+      consoleSpy.mockImplementation(() => {});
+      const codeGenSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      auth = new Auth({} as any);
+      auth.login();
+      expect(consoleSpy).toHaveBeenCalled();
+      expect(codeGenSpy).toHaveBeenCalledTimes(0);
+
+      consoleSpy.mockRestore();
+      codeGenSpy.mockRestore();
+    });
+
+    it('opens a window to the wally api login page with clientId and state', () => {
+      const state = 'mississippi';
+      const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
+
+      jest
+        .spyOn(Auth.prototype as any, 'generateStateCode')
+        .mockImplementation(() => state);
+
+      auth.login();
+
+      expect(storageSpy).toHaveBeenCalledWith(
+        `wally:${clientId}:state:token`,
+        state
+      );
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}`,
+        '_blank'
+      );
+
+      storageSpy.mockRestore();
+    });
+
+    it('passes redirectUrl and email when set', () => {
+      const state = 'mississippi';
+      const email = 'joe@smith.net';
+      const redirectURL = 'http://salmonofcapistrano.com';
+
+      const genCodeSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      genCodeSpy.mockImplementation(() => state);
+
+      auth = new Auth({
+        clientId,
+        host,
+        redirectURL,
+        messenger: new Messenger({}),
+      } as any);
+      auth.login(email);
+
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}&redirectUrl=${encodeURIComponent(
+          redirectURL
+        )}&email=${encodeURIComponent(email)}`,
+        '_blank'
+      );
+      genCodeSpy.mockRestore();
+    });
+
+    describe('returned promise', () => {
```

Comment:
Q: Is it possible to test that the events are triggered? It'll make the test stronger. At the same time, it doesn't have to test internal implementation details as long as the callback is triggered after some action.

Reply:
Alright, I made one test for the happy path all the way through the login flow, hopefully that's what you're talking about. I know testing implementation details isn't awesome 😅  but I'd like to leave those tests in to explicitly handle edge cases/error handling if that's alright